report/41_Lessons-Evolution-&-Refactoring.md

```diff
@@ -1 +1,97 @@
+
+Implementation of a proper test suite came later down the line, with
+The team was always aware of the fact that there should be more tests, and were even reminded when things broke in production (like when the register page broke completely).
+Though it is a lesson that the team has already learned in previous projects, here the team learned it once again.
+
+### Evolution
+
+Once the necessary refactoring steps were done, the team was capable of expanding the original implementation.
+
+#### Automation is king
+
+Early on in the development process, not many automatic steps were set up in the project pipeline.
+This made evolution much harder than it had to be, as tests would have to be manually run all the time to verify builds, and releases would have to be done manually.
+
+In the pipeline, the team has focused on having many actions done automatically, as mentioned in [TODO: 2.3: CI/CD](23_Process-CI-CD-Description.md).
+This was, at first, because the course demanded it, but once the team started using automatic steps, then the value was apparent.
+The pipeline was expanded repeatedly, with the steps themselves even being expanded, as is the case with `e2e.yml`.
+
+The takeaway here is that automation is very useful, as these manual, menial tasks could now be carried out automatically.
+A focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible.
+
+#### Choose scalable technologies early
+
+The team knew that choosing scalable techonologies early was important.
+If a choice to switch from a monolithic server to a distributed system comes too late, then the changes that will have to be made could become quite big, resulting in a lot of technical debt.
+
+This was why dockerization came early for the project, with separation between components for easier clusterization.
+The team didn't quite get Docker Swarm to work, but because these considerations were done early, Docker Swarm was actually obtainable, instead of being buried in debt from not having a RESTful API, for example.
+
+#### Graphs are motivational
+
+After the simulator had been live for a few weeks, the statistics showed that this project lacked behind the other participating projects by quite a big margin.
+Having this as a motivator to improve was vital to finding some of the bugs in the system.
+
+First, it was discovered that the project lacked behind because response times were highly influential for the graph trend lines.
+It was then considered whether HTTPS and Traefik routing were the thing slowing down the simulator, but removing those things only made very slight improvements, so they were re-enabled.
+The actual reason turned out to be the server location, as will be described in [TODO: 4.2 Operations](42_Lessons-Operation.md).
+
+Without the graph as a motivator, these things would never have been discovered.
```

Comment:
Arguably, this sections is not so much about graphs as it is about statistics, or likely, about competition. An interesting reflection would be how to recreate the feeling of competition in a real life environment, since this situation with 10 groups doing the same project would never happen in real life.